File: src/main/java/com/rkroom/blog/utility/JWTUtil.java
package com.rkroom.blog.utility;

import com.auth0.jwt.JWT;
import com.auth0.jwt.JWTVerifier;
import com.auth0.jwt.algorithms.Algorithm;
import com.auth0.jwt.exceptions.JWTDecodeException;
import com.auth0.jwt.interfaces.DecodedJWT;

import java.util.Date;

public class JWTUtil {
    // 过期时间七天
    private static final long EXPIRE_TIME = 7 * 24 * 60 * 60 * 1000;

    /*
     * 生成签名,过期时间为EXPIRE_TIME，将密码hash之后用作签名
     * @param username 用户名
     * @param secret 用户的密码
     * @return 加密的token
     */
    // todo 将密码加盐再作为密钥
    public static String sign(String username, int userId, String secret) {
        try {
            Date date = new Date(System.currentTimeMillis() + EXPIRE_TIME);
            Algorithm algorithm = Algorithm.HMAC256(secret);
            // 附带username和userId信息
            return JWT.create()
                    .withClaim("username", username)
                    .withClaim("id", userId)
                    .withExpiresAt(date)
                    .sign(algorithm);
        } catch (Exception e) {
            return null;
        }
    }

    /*
     * 校验token是否正确
     * @param token 密钥
     * @param secret 用户的密码
     * @return 是否正确
     */
    public static boolean verify(String token, String username, String secret) {
        try {
            Algorithm algorithm = Algorithm.HMAC256(secret);
            // 解析并验证token
            JWTVerifier verifier = JWT.require(algorithm)
                    .withClaim("username", username)
                    .build();
            DecodedJWT jwt = verifier.verify(token);
            return true;
        } catch (Exception exception) {
            return false;
        }
    }

    /*
     * 获得token中的信息，无需secret解密也能获得
     * @return token中包含的用户名
     */
    public static String getUsername(String token) {
        try {
            DecodedJWT jwt = JWT.decode(token);
            return jwt.getClaim("username").asString();
        } catch (JWTDecodeException e) {
            return null;
        }
    }
}

File: src/main/java/com/rkroom/blog/utility/JWTUtil.kt
package com.rkroom.blog.utility

import com.auth0.jwt.JWT
import com.auth0.jwt.algorithms.Algorithm
import com.auth0.jwt.exceptions.JWTDecodeException
import java.util.*

object JWTUtil {
    // 过期时间七天
    private const val EXPIRE_TIME = 7 * 24 * 60 * 60 * 1000.toLong()

    /*
     * 生成签名,过期时间为EXPIRE_TIME，将密码hash之后用作签名
     * @param username 用户名
     * @param secret 用户的密码
     * @return 加密的token
     */
    fun sign(username: String?, userId: Int, secret: String?): String? {
        return try {
            val date = Date(System.currentTimeMillis() + EXPIRE_TIME)
            val algorithm = Algorithm.HMAC256(secret)
            // 附带username和userId信息
            JWT.create()
                    .withClaim("username", username)
                    .withClaim("id", userId)
                    .withExpiresAt(date)
                    .sign(algorithm)
        } catch (e: Exception) {
            null
        }
    }

    /*
     * 校验token是否正确
     * @param token 密钥
     * @param secret 用户的密码
     * @return 是否正确
     */
    @JvmStatic
    fun verify(token: String?, username: String?, secret: String?): Boolean {
        return try {
            val algorithm = Algorithm.HMAC256(secret)
            // 解析并验证token
            val verifier = JWT.require(algorithm)
                    .withClaim("username", username)
                    .build()
            val jwt = verifier.verify(token)
            true
        } catch (exception: Exception) {
            false
        }
    }

    /*
     * 获得token中的信息，无需secret解密也能获得
     * @return token中包含的用户名
     */
    @JvmStatic
    fun getUsername(token: String?): String? {
        return try {
            val jwt = JWT.decode(token)
            jwt.getClaim("username").asString()
        } catch (e: JWTDecodeException) {
            null
        }
    }
}